hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/placement/policy/MultiComparatorPolicy.java

```diff
@@ -0,0 +1,346 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.CapacitySchedulerConfiguration.DOT;
+
+/**
+ * <p>
+ * This class has the following functionality:
+ *
+ * <p>
+ * MultiComparatorPolicy
+ * - manages some common comparators to help sorting nodes by
+ *      allocated/unallocated/total resource, dominant ratio, etc.
+ * - holds sorted nodes list based on the of nodes at given time.
+ * - can be configured with specified comparators.
+ * </p>
+ */
+public class MultiComparatorPolicy<N extends SchedulerNode>
+    implements MultiNodeLookupPolicy<N>, Configurable {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(MultiComparatorPolicy.class);
+  // comparators
+  private static final DominantResourceCalculator DOMINANT_RC =
+      new DominantResourceCalculator();
+  private static final Map<ComparatorKey, Function<SchedulerNode, Comparable>>
+      COMPARATOR_CALCULATORS = Collections.unmodifiableMap(
+      new HashMap<ComparatorKey, Function<SchedulerNode, Comparable>>() {{
+        // for vcores
+        put(ComparatorKey.ALLOCATED_VCORES,
+            obj -> obj.getAllocatedResource().getVirtualCores());
+        put(ComparatorKey.UNALLOCATED_VCORES,
+            obj -> obj.getUnallocatedResource().getVirtualCores());
+        put(ComparatorKey.TOTAL_VCORES,
+            obj -> obj.getTotalResource().getVirtualCores());
+        // for memory
+        put(ComparatorKey.ALLOCATED_MEMORY,
+            obj -> obj.getAllocatedResource().getMemorySize());
+        put(ComparatorKey.UNALLOCATED_MEMORY,
+            obj -> obj.getUnallocatedResource().getMemorySize());
+        put(ComparatorKey.TOTAL_MEMORY,
+            obj -> obj.getTotalResource().getMemorySize());
+        // for resource
+        put(ComparatorKey.ALLOCATED_RESOURCE,
+            SchedulerNode::getAllocatedResource);
+        put(ComparatorKey.UNALLOCATED_RESOURCE,
+            SchedulerNode::getUnallocatedResource);
+        put(ComparatorKey.TOTAL_RESOURCE,
+            SchedulerNode::getTotalResource);
+        // for dominant ratio
+        put(ComparatorKey.DOMINANT_ALLOCATED_RATIO, obj -> Resources
+            .ratio(DOMINANT_RC, obj.getAllocatedResource(),
+                obj.getTotalResource()));
+        // for node ID
+        put(ComparatorKey.NODE_ID, SchedulerNode::getNodeID);
+      }});
+  // conf keys and default values
+  public static final String COMPARATORS_CONF_KEY = "comparators";
+  protected static final List<Comparator> DEFAULT_COMPARATORS = Collections
+      .unmodifiableList(Arrays.asList(
```

Comment:
So this means the default sorting policy is based on resource utilization, the node having less resource utilization will be given priority

Reply:
Yes, default comparators will only be used when there's no comparators configured for this policy or configured incorrectly. Because I believe optimizing the workload distribution among nodes is the primary use case.